### ISSUE TITLE
netdisco-do discover race condition

### ISSUE THREAD

**Issue by inphobia** [netdisco/netdisco#671]
## Expected Behavior

being able to run netdisco-do discover on multiple devices with 1 command.

## Current Behavior

at least for interconnected systems this has produced a race condition.

this seems to happen only for unknown devices on initial discovery, once they are known i am no longer able to reproduce this issue.

## Possible Solution

at first glance this seems like a race condition between checking if a device exists and doing the actual insert in the db. 
* perhaps netdisco-do discover needs to postpone neighbor discovery until it's done discovering the devices defined on the cli
* perhaps we need more database locks
* perhaps multiple devices should imply --enqueue
* perhaps you just trust netdisco to discover neighbors

all joking aside, i'm not sure how to fix this and few people will ever run into it i guess. perhaps someone has a better insight in the caching & scheduling can come up with an idea.

the actual impact is almost non existant. all 4 devices were discovered correctly, but since the last 2 on the commandline were discovered via neighbor discovery the also already ran arpnip & macsuck, while the first 2 devices will need to wait for the arp/macwalk schedule to get their first run.

## Steps to Reproduce (for bugs)

1. clean netdisco setup (no devices in db yet)
2. 4 nexus switches, set up as vpc pairs, interconnected over multiple links
3. `netdisco-do discover -d 10.1.100.250 -d 10.1.100.251 -d 10.1.100.252 -d 10.1.100.253`

```
[2995] 2019-10-30 20:25:06  info App::Netdisco version 2.044004 loaded.
[2995] 2019-10-30 20:25:06  info discover: [10.1.100.250] started at Wed Oct 30 21:25:06 2019
[2995] 2019-10-30 20:25:10  info discover: finished at Wed Oct 30 21:25:10 2019
[2995] 2019-10-30 20:25:10  info discover: status done: Ended discover for 10.1.100.250
[2995] 2019-10-30 20:25:10  info discover: [10.1.100.251] started at Wed Oct 30 21:25:10 2019
[2995] 2019-10-30 20:25:14  info discover: finished at Wed Oct 30 21:25:14 2019
[2995] 2019-10-30 20:25:14  info discover: status done: Ended discover for 10.1.100.251
[2995] 2019-10-30 20:25:14  info discover: [10.1.100.252] started at Wed Oct 30 21:25:14 2019
[2995] 2019-10-30 20:25:15  info discover: finished at Wed Oct 30 21:25:15 2019
[2995] 2019-10-30 20:25:15  info discover: status error: DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR:  duplicate key value violates unique constraint "device_pkey"
DETAIL:  Key (ip)=(10.1.100.252) already exists. [for Statement "INSERT INTO device ( contact, description, fan, ip, last_discover, layers, location, mac, model, name, os, os_ver, ports, ps1_status, ps1_type, ps2_status, ps2_type, serial, slots, snmp_class, snmp_ver, uptime, vendor, vtp_domain) VALUES ( ?, ?, ?, ?, now(), ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ? )" with ParamValues: 1='systeembeheer', 2='Cisco NX-OS(tm) nxos.7.0.3.I7.7.bin, Software (nxos), Version 7.0(3)I7(7), RELEASE SOFTWARE Copyright (c) 2002-2019 by Cisco Systems, Inc. Compiled 8/28/2019 15:00:00', 3=undef, 4='10.1.100.252', 5='01000110', 6='d10', 7='00:23:04:ee:be:0a', 8='N9KC9332PQ', 9='d10-n9k1.aquafin.be', 10='nx-os', 11='7.0(3)I7(7)', 12='123', 13=undef, 14=undef, 15=undef, 16=undef, 17='FDO21392TG5', 18=undef, 19='SNMP::Info::Layer3::Nexus', 20='3', 21='402102652', 22='cisco', 23=''] at /home/testdisc/perl5/lib/perl5/App/Netdisco/Worker/Plugin/Discover/Properties.pm line 58
[2995] 2019-10-30 20:25:15  info discover: [10.1.100.253] started at Wed Oct 30 21:25:15 2019
[2995] 2019-10-30 20:25:15  info discover: finished at Wed Oct 30 21:25:15 2019
[2995] 2019-10-30 20:25:15  info discover: status error: DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: ERROR:  duplicate key value violates unique constraint "device_pkey"
DETAIL:  Key (ip)=(10.1.100.253) already exists. [for Statement "INSERT INTO device ( contact, description, fan, ip, last_discover, layers, location, mac, model, name, os, os_ver, ports, ps1_status, ps1_type, ps2_status, ps2_type, serial, slots, snmp_class, snmp_ver, uptime, vendor, vtp_domain) VALUES ( ?, ?, ?, ?, now(), ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ? )" with ParamValues: 1='systeembeheer', 2='Cisco NX-OS(tm) nxos.7.0.3.I7.7.bin, Software (nxos), Version 7.0(3)I7(7), RELEASE SOFTWARE Copyright (c) 2002-2019 by Cisco Systems, Inc. Compiled 8/28/2019 15:00:00', 3=undef, 4='10.1.100.253', 5='01000110', 6='d10', 7='00:23:04:ee:be:0a', 8='N9KC9332PQ', 9='d10-n9k2.aquafin.be', 10='nx-os', 11='7.0(3)I7(7)', 12='123', 13=undef, 14=undef, 15=undef, 16=undef, 17='FDO21392TFD', 18=undef, 19='SNMP::Info::Layer3::Nexus', 20='3', 21='401990794', 22='cisco', 23=''] at /home/testdisc/perl5/lib/perl5/App/Netdisco/Worker/Plugin/Discover/Properties.pm line 58
```

## Context

after cleaning out my test setup i wanted to add a few devices to test again, so i used the new feature to define multiple targets at once. since they are all interconnected and running cdp & lldp i'm guessing after the first device is discovered it fires up it's neighbor discovery so the switches defined last on the commandline are already in the process of being discovered when netdisco-do gets to them.

## Your Environment

* Netdisco version used: 2.044004
* SNMP::Info version used: 3.70

**Comment by ollyg** [comment]
Interesting note that the first two devices didn't get arp/mac gathered. They should have been run with the `with-nodes` option to do this. So there's something to fix in `netdisco-do` at least!

**Comment by inphobia** [comment]
> Interesting note that the first two devices didn't get arp/mac gathered. They should have been run with the `with-nodes` option to do this.

iirc netdisco-do just runs the selected action, discover in this case (discovered neighbors however will be run with discover with-nodes). arpnic/macwalk will be run when the arpwalk/macwalk schedule trigger. i tried just about all options i could think of but could not get netdisco-do to access the `with-nodes` parameter.

**Comment by ollyg** [comment]
fixed in [master 92875f0e] fix #671 need to discover nodes on CLI discover
